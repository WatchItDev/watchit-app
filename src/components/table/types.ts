// ----------------------------------------------------------------------

export type TableProps = {
  dense: boolean;
  page: number;
  rowsPerPage: number;
  order: 'asc' | 'desc';
  orderBy: string;
  //
  selected: string[];
  onSelectRow: (id: string) => void;
  onSelectAllRows: (checked: boolean, newSelecteds: string[]) => void;
  //
  onResetPage: VoidFunction;
  onSort: (id: string) => void;
  onChangePage: (event: unknown, newPage: number) => void;
  onChangeRowsPerPage: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onChangeDense: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onUpdatePageDeleteRow: (totalRowsInPage: number) => void;
  onUpdatePageDeleteRows: ({
    totalRows,
    totalRowsInPage,
    totalRowsFiltered,
  }: {
    totalRows: number;
    totalRowsInPage: number;
    totalRowsFiltered: number;
  }) => void;
  //
  setPage: React.Dispatch<React.SetStateAction<number>>;
  setDense: React.Dispatch<React.SetStateAction<boolean>>;
  setOrder: React.Dispatch<React.SetStateAction<'desc' | 'asc'>>;
  setOrderBy: React.Dispatch<React.SetStateAction<string>>;
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
  setRowsPerPage: React.Dispatch<React.SetStateAction<number>>;
};
