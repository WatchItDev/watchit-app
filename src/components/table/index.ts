export * from './types';
export * from './utils';

export { default as useTable } from './use-table';

export { default as TableNoData } from './table-no-data';
export { default as TableSkeleton } from './table-skeleton';
export { default as TableEmptyRows } from './table-empty-rows';
export { default as TableHeadCustom } from './table-head-custom';
export { default as TableSelectedAction } from './table-selected-action';
export { default as TablePaginationCustom } from './table-pagination-custom';
