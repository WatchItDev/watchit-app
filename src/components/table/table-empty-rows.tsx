// @mui
import TableRow from '@mui/material/TableRow';
import TableCell from '@mui/material/TableCell';

// ----------------------------------------------------------------------

type Props = {
  height?: number;
  emptyRows: number;
};

export default function TableEmptyRows({ emptyRows, height }: Props) {
  if (!emptyRows) {
    return null;
  }

  return (
    <TableRow
      sx={{
        ...(height && {
          height: height * emptyRows,
        }),
      }}
    >
      <TableCell colSpan={9} />
    </TableRow>
  );
}
