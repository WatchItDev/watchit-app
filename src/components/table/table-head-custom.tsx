// @mui
import { Theme, SxProps } from '@mui/material/styles';
import Box from '@mui/material/Box';
import TableRow from '@mui/material/TableRow';
import Checkbox from '@mui/material/Checkbox';
import TableHead from '@mui/material/TableHead';
import TableCell from '@mui/material/TableCell';
import TableSortLabel from '@mui/material/TableSortLabel';

// ----------------------------------------------------------------------

const visuallyHidden = {
  border: 0,
  margin: -1,
  padding: 0,
  width: '1px',
  height: '1px',
  overflow: 'hidden',
  position: 'absolute',
  whiteSpace: 'nowrap',
  clip: 'rect(0 0 0 0)',
} as const;

// ----------------------------------------------------------------------

type Props = {
  order?: 'asc' | 'desc';
  orderBy?: string;
  headLabel: any[];
  rowCount?: number;
  numSelected?: number;
  onSort?: (id: string) => void;
  onSelectAllRows?: (checked: boolean) => void;
  sx?: SxProps<Theme>;
};

export default function TableHeadCustom({
  order,
  orderBy,
  rowCount = 0,
  headLabel,
  numSelected = 0,
  onSort,
  onSelectAllRows,
  sx,
}: Props) {
  return (
    <TableHead sx={sx}>
      <TableRow>
        {onSelectAllRows && (
          <TableCell padding="checkbox">
            <Checkbox
              indeterminate={!!numSelected && numSelected < rowCount}
              checked={!!rowCount && numSelected === rowCount}
              onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
                onSelectAllRows(event.target.checked)
              }
            />
          </TableCell>
        )}

        {headLabel.map((headCell) => (
          <TableCell
            key={headCell.id}
            align={headCell.align || 'left'}
            sortDirection={orderBy === headCell.id ? order : false}
            sx={{ width: headCell.width, minWidth: headCell.minWidth }}
          >
            {onSort ? (
              <TableSortLabel
                hideSortIcon
                active={orderBy === headCell.id}
                direction={orderBy === headCell.id ? order : 'asc'}
                onClick={() => onSort(headCell.id)}
              >
                {headCell.label}

                {orderBy === headCell.id ? (
                  <Box sx={{ ...visuallyHidden }}>
                    {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
                  </Box>
                ) : null}
              </TableSortLabel>
            ) : (
              headCell.label
            )}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
}
