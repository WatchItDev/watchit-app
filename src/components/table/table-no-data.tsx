// @mui
import { Theme, SxProps } from '@mui/material/styles';
import TableRow from '@mui/material/TableRow';
import TableCell from '@mui/material/TableCell';
//
import EmptyContent from '../empty-content';

// ----------------------------------------------------------------------

type Props = {
  notFound: boolean;
  sx?: SxProps<Theme>;
};

export default function TableNoData({ notFound, sx }: Props) {
  return (
    <TableRow>
      {notFound ? (
        <TableCell colSpan={12}>
          <EmptyContent
            filled
            title="No Data"
            sx={{
              py: 10,
              ...sx,
            }}
          />
        </TableCell>
      ) : (
        <TableCell colSpan={12} sx={{ p: 0 }} />
      )}
    </TableRow>
  );
}
