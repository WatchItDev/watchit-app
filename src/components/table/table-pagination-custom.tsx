// @mui
import { Theme, SxProps } from '@mui/material/styles';
import Box from '@mui/material/Box';
import Switch from '@mui/material/Switch';
import FormControlLabel from '@mui/material/FormControlLabel';
import TablePagination, { TablePaginationProps } from '@mui/material/TablePagination';

// ----------------------------------------------------------------------

type Props = {
  dense?: boolean;
  onChangeDense?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  sx?: SxProps<Theme>;
};

export default function TablePaginationCustom({
  dense,
  onChangeDense,
  rowsPerPageOptions = [5, 10, 25],
  sx,
  ...other
}: Props & TablePaginationProps) {
  return (
    <Box sx={{ position: 'relative', ...sx }}>
      <TablePagination
        rowsPerPageOptions={rowsPerPageOptions}
        component="div"
        {...other}
        sx={{
          borderTopColor: 'transparent',
        }}
      />

      {onChangeDense && (
        <FormControlLabel
          label="Dense"
          control={<Switch checked={dense} onChange={onChangeDense} />}
          sx={{
            pl: 2,
            py: 1.5,
            top: 0,
            position: {
              sm: 'absolute',
            },
          }}
        />
      )}
    </Box>
  );
}
