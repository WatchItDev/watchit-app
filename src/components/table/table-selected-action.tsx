// @mui
import Checkbox from '@mui/material/Checkbox';
import Typography from '@mui/material/Typography';
import Stack, { StackProps } from '@mui/material/Stack';

// ----------------------------------------------------------------------

interface Props extends StackProps {
  dense?: boolean;
  action?: React.ReactNode;
  rowCount: number;
  numSelected: number;
  onSelectAllRows: (checked: boolean) => void;
}

export default function TableSelectedAction({
  dense,
  action,
  rowCount,
  numSelected,
  onSelectAllRows,
  sx,
  ...other
}: Props) {
  if (!numSelected) {
    return null;
  }

  return (
    <Stack
      direction="row"
      alignItems="center"
      sx={{
        pl: 1,
        pr: 2,
        top: 0,
        left: 0,
        width: 1,
        zIndex: 9,
        height: 58,
        position: 'absolute',
        bgcolor: 'primary.lighter',
        ...(dense && {
          height: 38,
        }),
        ...sx,
      }}
      {...other}
    >
      <Checkbox
        indeterminate={!!numSelected && numSelected < rowCount}
        checked={!!rowCount && numSelected === rowCount}
        onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
          onSelectAllRows(event.target.checked)
        }
      />

      <Typography
        variant="subtitle2"
        sx={{
          ml: 2,
          flexGrow: 1,
          color: 'primary.main',
          ...(dense && {
            ml: 3,
          }),
        }}
      >
        {numSelected} selected
      </Typography>

      {action && action}
    </Stack>
  );
}
