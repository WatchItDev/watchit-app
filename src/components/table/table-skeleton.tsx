// @mui
import Stack from '@mui/material/Stack';
import Skeleton from '@mui/material/Skeleton';
import TableCell from '@mui/material/TableCell';
import TableRow, { TableRowProps } from '@mui/material/TableRow';

// ----------------------------------------------------------------------

export default function TableSkeleton({ ...other }: TableRowProps) {
  return (
    <TableRow {...other}>
      <TableCell colSpan={12}>
        <Stack spacing={3} direction="row" alignItems="center">
          <Skeleton sx={{ borderRadius: 1.5, width: 48, height: 48, flexShrink: 0 }} />
          <Skeleton sx={{ width: 1, height: 12 }} />
          <Skeleton sx={{ width: 180, height: 12 }} />
          <Skeleton sx={{ width: 160, height: 12 }} />
          <Skeleton sx={{ width: 140, height: 12 }} />
          <Skeleton sx={{ width: 120, height: 12 }} />
        </Stack>
      </TableCell>
    </TableRow>
  );
}
