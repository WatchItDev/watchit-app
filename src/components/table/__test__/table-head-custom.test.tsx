import { fireEvent, render, screen } from '@testing-library/react';
import TableHeadCustom from '../table-head-custom';
import { HeadLabel } from '@src/components/table';

describe('[COMPONENTS]: TableHeadCustom component testing', () => {
  const mockHeadLabels: HeadLabel[] = [
    { id: 'name', label: 'Name' },
    { id: 'email', label: 'Email', align: 'center', width: 200 },
    { id: 'role', label: 'Role', align: 'right', minWidth: 150 },
  ];

  it('matches snapshot', () => {
    const { container } = render(
      <TableHeadCustom headLabel={mockHeadLabels} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with column headers correctly', () => {
    render(<TableHeadCustom headLabel={mockHeadLabels} />);

    expect(screen.getByText('Name')).toBeInTheDocument();
    expect(screen.getByText('Email')).toBeInTheDocument();
    expect(screen.getByText('Role')).toBeInTheDocument();
  });

  it('renders without checkbox when onSelectAllRows is not provided', () => {
    const { container } = render(
      <TableHeadCustom headLabel={mockHeadLabels} />,
    );

    const checkboxes = container.querySelectorAll('input[type="checkbox"]');
    expect(checkboxes.length).toBe(0);
  });

  it('renders with checkbox when onSelectAllRows is provided', () => {
    const mockSelectAllRows = vi.fn();

    render(
      <TableHeadCustom
        headLabel={mockHeadLabels}
        onSelectAllRows={mockSelectAllRows}
        rowCount={5}
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeInTheDocument();
  });

  it('calls onSelectAllRows when checkbox is clicked', async () => {
    const mockSelectAllRows = vi.fn();

    render(
      <TableHeadCustom
        headLabel={mockHeadLabels}
        onSelectAllRows={mockSelectAllRows}
        rowCount={5}
      />,
    );

    const checkbox = screen.getByRole('checkbox');

    fireEvent.click(checkbox);

    expect(mockSelectAllRows).toHaveBeenCalledWith(true);
  });

  it('shows checked checkbox when all rows are selected', () => {
    const mockSelectAllRows = vi.fn();

    render(
      <TableHeadCustom
        headLabel={mockHeadLabels}
        onSelectAllRows={mockSelectAllRows}
        rowCount={5}
        numSelected={5}
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeChecked();
  });

  it('renders sort labels when onSort is provided', () => {
    const mockOnSort = vi.fn();

    render(<TableHeadCustom headLabel={mockHeadLabels} onSort={mockOnSort} />);

    const sortLabels = screen.getAllByRole('button');
    expect(sortLabels).toHaveLength(3);
  });

  it('displays sort direction text for screen readers', () => {
    const mockOnSort = vi.fn();

    render(
      <TableHeadCustom
        headLabel={mockHeadLabels}
        onSort={mockOnSort}
        order="asc"
        orderBy="name"
      />,
    );

    expect(screen.getByText('sorted ascending')).toBeInTheDocument();
  });

  it('handles zero row count correctly', () => {
    const mockSelectAllRows = vi.fn();

    render(
      <TableHeadCustom
        headLabel={mockHeadLabels}
        onSelectAllRows={mockSelectAllRows}
        rowCount={0}
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
  });

  it('applies custom sx to TableHead', () => {
    const customSx = { backgroundColor: 'rgb(240, 240, 240)' };

    const { container } = render(
      <TableHeadCustom headLabel={mockHeadLabels} sx={customSx} />,
    );

    const tableHead = container.querySelector('thead');
    expect(tableHead).toHaveStyle('background-color: rgb(240, 240, 240)');
  });
});
