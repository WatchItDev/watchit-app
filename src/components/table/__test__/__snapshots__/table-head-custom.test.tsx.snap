// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[COMPONENTS]: TableHeadCustom component testing > matches snapshot 1`] = `
<div>
  <thead
    class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
        scope="col"
      >
        Name
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-elqmqc-MuiTableCell-root"
        scope="col"
      >
        Email
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-10wjn0w-MuiTableCell-root"
        scope="col"
      >
        Role
      </th>
    </tr>
  </thead>
</div>
`;
